vm/src/stdlib/socket.rs

```diff
@@ -1,493 +1,317 @@
-        vm: &VirtualMachine,
-    ) -> PyResult<SocketRef> {
-        Socket::new(family, kind).into_ref_with_type(vm, cls)
-    }
-
-    fn enter(self, _vm: &VirtualMachine) -> SocketRef {
-        self
-    }
-
-    fn exit(self, _args: PyFuncArgs, _vm: &VirtualMachine) {
-        self.close(_vm)
-    }
-
-    fn connect(self, address: Address, vm: &VirtualMachine) -> PyResult<()> {
-        let address_string = address.get_address_string();
-
-        match self.socket_kind {
-            SocketKind::Stream => {
-                let con = if let Some(duration) = self.timeout.borrow().as_ref() {
-                    let sock_addr = match address_string.to_socket_addrs() {
-                        Ok(mut sock_addrs) => {
-                            if sock_addrs.len() == 0 {
-                                let error_type = vm.class("socket", "gaierror");
-                                return Err(vm.new_exception(
-                                    error_type,
-                                    "nodename nor servname provided, or not known".to_string(),
-                                ));
-                            } else {
-                                sock_addrs.next().unwrap()
-                            }
-                        }
-                        Err(e) => {
-                            let error_type = vm.class("socket", "gaierror");
-                            return Err(vm.new_exception(error_type, e.to_string()));
-                        }
-                    };
-                    TcpStream::connect_timeout(&sock_addr, *duration)
-                } else {
-                    TcpStream::connect(address_string)
-                };
-                match con {
-                    Ok(stream) => {
-                        self.con.borrow_mut().replace(Connection::TcpStream(stream));
-                        Ok(())
-                    }
-                    Err(ref e) if e.kind() == io::ErrorKind::TimedOut => {
-                        let socket_timeout = vm.class("socket", "timeout");
-                        Err(vm.new_exception(socket_timeout, "Timed out".to_string()))
-                    }
-                    Err(s) => Err(vm.new_os_error(s.to_string())),
-                }
-            }
-            SocketKind::Dgram => {
-                if let Some(Connection::UdpSocket(con)) = self.con.borrow().as_ref() {
-                    match con.connect(address_string) {
-                        Ok(_) => Ok(()),
-                        Err(s) => Err(vm.new_os_error(s.to_string())),
-                    }
-                } else {
-                    Err(vm.new_type_error("".to_string()))
-                }
+    #[pymethod]
+    fn recvfrom(&self, bufsize: usize, vm: &VirtualMachine) -> PyResult {
+        let mut buffer = vec![0u8; bufsize];
+        let addr = match self.sock().recv_from(&mut buffer) {
+            Ok((_, addr)) => addr,
+            Err(ref e) if e.kind() == io::ErrorKind::TimedOut => {
```

Comment:
I would suggest adding a function name `convert_socket_error` that will check if this is a timeout and then call `convert_io_error`. That will remove a lot of duplication.